map/framework.cpp

```diff
@@ -3310,3 +3310,73 @@ void Framework::OnPowerSchemeChanged(power_management::Scheme const actualScheme
   if (actualScheme == power_management::Scheme::EconomyMaximum && GetTrafficManager().IsEnabled())
     GetTrafficManager().SetEnabled(false);
 }
+
+bool Framework::ShouldShowProducts() const
+{
+  auto const connectionStatus = Platform::ConnectionStatus();
+  if (connectionStatus == Platform::EConnectionType::CONNECTION_NONE)
+    return false;
+
+  std::string donateUrl;
+  if (!settings::Get(settings::kDonateUrl, donateUrl)) // donation is disabled
+    return false;
+
+  if (!storage::IsPointCoveredByDownloadedMaps(GetCurrentPlacePageInfo().GetMercator(), m_storage, * m_infoGetter))
+    return false;
+
+  static uint32_t kFirstLaunchTimeout;
+  static uint32_t kLastProductCloseTimeout;
+  static uint32_t kLastProductSelectTimeout;
+  #ifdef DEBUG
+  kFirstLaunchTimeout = 60 * 3; // 3 min because it may takes some time to download the map after the first launch in debug
+  kLastProductCloseTimeout = 10;
+  kLastProductSelectTimeout = 20;
+  #else
+  kFirstLaunchTimeout = 60 * 60 * 3; // 3 hours
+  kLastProductCloseTimeout = 60 * 60 * 24  * 14; // 14 days
+  kLastProductSelectTimeout = 180 * 24 * 60 * 60; // 180 days
+  #endif
+
+  auto const now = base::SecondsSinceEpoch();
+
+  if (m_usageStats.GetFirstLaunchTime() + kFirstLaunchTimeout > now)
+    return false;
+
+  uint64_t popupCloseTime;
+  if (!settings::Get(settings::kPlacePageProductsCloseTime, popupCloseTime))
+    settings::Set(settings::kPlacePageProductsCloseTime, 0);
+  uint64_t productSelectTime;
+  if (!settings::Get(settings::kPlacePageProductsSelectedTime, productSelectTime))
+    settings::Set(settings::kPlacePageProductsSelectedTime, 0);
+
+  bool const closePopupTimeoutExpired = popupCloseTime == 0 || popupCloseTime + kLastProductCloseTimeout < now;
+  bool const productSelectTimeoutExpired = productSelectTime == 0 || productSelectTime + kLastProductSelectTimeout < now;
+  if (closePopupTimeoutExpired && productSelectTimeoutExpired)
+    return true;
+
+  return false;
+}
+
+std::optional<products::ProductsConfig> Framework::GetProductsConfiguration() const
+{
+  if (!ShouldShowProducts())
+    return nullopt;
+  return products::GetProductsConfiguration();
```

Comment:
ProductSettings::Instance().Get()

We can store actual already parsed instance do avoid _parsing_ every time we show PP:

```
class ProductSettings
{
public:
  void Update(std::string const & json)
  {
    lock_guard guard(m_mutex);
    FileWriter w(...);
    w.Write(json);
    m_config = Parse(json);
  }
  ProductConfig Get()
  {
    lock_guard guard(m_mutex);
    return m_config;
  }
};

```

Reply:
Fixed! Thanks